### PR TITLE
PlayerControllerSystem player and editor control component separation

### PR DESCRIPTION
**Pull request by Kotailri**
### Description
Separated PlayerControllerComponent into PlayerControllerComponent and EditorControllerComponent and made the PlayerControllerSystem query entities with either of those components to run their specific movement operations.

### Trello Ticket
https://trello.com/c/7gEPHOtP/48-make-a-query-in-playercontroller-for-editorcontrollercomponents

### Additional Info
PlayerController was renamed to PlayerControllerSystem and moved into the ecs systems folder, so all the PlayerControllerSystem includes were also modified.

### CODE REVIEW

File: Code/Engine/ECS/Systems/PlayerControllerSystem.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#include <Engine/ECS/Components/TransformComponent.h>
+#include <Engine/InputManager.h>
+
+namespace sad
+{
+	class PlayerControllerSystem
+	{
+	public:
+		PlayerControllerSystem();
+		~PlayerControllerSystem();
```
Comment (jake-pauls) [inline Code/Engine/ECS/Systems/PlayerControllerSystem.h]: Do we need to have constructors and destructors defined if all methods on the system are `static`?

File: Code/Engine/ECS/Systems/PlayerControllerSystem.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+		static void Update();
```
Comment (jake-pauls) [inline Code/Engine/ECS/Systems/PlayerControllerSystem.h]: What do you think about passing a reference to the `EntityWorld&` in our system updates?

Do you know if it's better to pass by reference (and have a reference in the system) or call `ecs::Registry::GetEntityWorld()` each time we want the world in each method of the system?

File: Code/Engine/ECS/Systems/PlayerControllerSystem.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		std::string mousePosition = "Mouse Pos: (";
+		mousePosition += std::to_string(input.GetMousePosition().x);
+		mousePosition += ", ";
+		mousePosition += std::to_string(input.GetMousePosition().y);
+		mousePosition += ")";
+
+		core::Log(ELogType::Info, mousePosition.c_str());
+	}
+
+	float movespeed = 0.005f;
```
Comment (jake-pauls) [inline Code/Engine/ECS/Systems/PlayerControllerSystem.cpp]: Prefer lowerCamelCase :peepobet:

File: Code/Engine/ECS/Systems/PlayerControllerSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+	sad::ecs::EntityWorld& world = sad::ecs::Registry::GetEntityWorld();
+	auto playerView = world.view<const sad::ecs::PlayerControllerComponent, const sad::ecs::TransformComponent>();
+	auto editorView = world.view<const sad::ecs::EditorControllerComponent, const sad::ecs::TransformComponent>();
```
Comment (jake-pauls) [inline Code/Engine/ECS/Systems/PlayerControllerSystem.cpp]: Not too sure about any performance differences here, but we might be able to make use of a cool entt feature called `exclude`.

In this case, we make the assumption that only one of these controllers should exist at a time - so we can exclude the opposite component in both view query (ie: exclude the `EditorControllerComponent` in our search for the `PlayerControllerComponent`). Might also be good just to be really explicit with the fact that their existence is mutual so we don't accidentally update both at once 😵 . 

An example from the entt crash course!

image

File: Code/Engine/ECS/Systems/PlayerControllerSystem.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+#include <Engine/ECS/Components/TransformComponent.h>
+#include <Engine/ECS/Components/PlayerControllerComponent.h>
+#include <Engine/ECS/Components/EditorControllerComponent.h>
+
+sad::PlayerControllerSystem::PlayerControllerSystem() {}
+sad::PlayerControllerSystem::~PlayerControllerSystem() {}
+
+void sad::PlayerControllerSystem::PlayerControls(InputManager& input, const ecs::TransformComponent& transformComponent, float movespeed)
+{
+	bool usingController = false; // bool that disables movement keys when joystick being used to prevent double speed
```
Comment (jake-pauls) [inline Code/Engine/ECS/Systems/PlayerControllerSystem.cpp]: Does this have to be determined by the system? Could we query this from the input manager - where we just toggle it whenever a controller/keyboard event is detected? 

For instance...
```
if (input.IsControllerActive())
{
    // Check controller inputs
}
else
{
    // Check other inputs
}
```

We already use an `m_ControllerIsActive` boolean in the InputManager which could be leveraged for this... (we could check for duplicate controllers in the `InputManager::OnControllerConnected` by checking `if (m_Controller)` instead)
```
void CatchControllerEvents(...)
{
    if (sdl detects controller event)
    {
        m_ControllerIsActive = true;
    }
}

void CatchKeyboardEvents(...)
{
    if (sdl detects keyboard event)
    {
        m_ControllerIsActive = false;
    }
}

bool IsControllerActive() { return m_ControllerIsActive; }
```

File: Code/Engine/ECS/Components/EditorControllerComponent.h [hunk 6]
```diff
@@ -0,0 +1,6 @@
+#pragma once
+
+namespace sad::ecs
+{
+	struct EditorControllerComponent {};
+}
```
Comment (jake-pauls) [inline Code/Engine/ECS/Components/EditorControllerComponent.h]: I remember I mentioned separating components in files 😅 but now I'm wondering if it makes sense to combine the editor and player controller components in one file? I'm cool with joining components in the same `.h` that are manipulated by the same ECS system - just curious what you think.

Also, would probably good to add an `@brief` just explaining what the component does 😄

**Review by jake-pauls — CHANGES_REQUESTED** [review]
Nice! Excited this is on the way - helps a lot with nailing more editor code this weekend. 🚀 

Just added some comments for things that may tighten this up a bit - also some comments are questions for preferences on how we handle some things going forward. Let me know what you think!

**Comment by jake-pauls** [comment]
[Make a query in PlayerController for EditorControllerComponent(s)](https://trello.com/c/7gEPHOtP/48-make-a-query-in-playercontroller-for-editorcontrollercomponents)

**Review by Kotailri — COMMENTED** [review]
Oops i mistyped in a commit, its "conflict from using keyboard and ~~keyboard~~ controller"

**Review by jake-pauls — APPROVED** [review]
Looking good! 🚀

**Review by paulc100 — APPROVED** [review]

**Pull request merged**